docs/dev/hacking-guide.md

```diff
@@ -0,0 +1,224 @@
+
+- Provider specific implementation. Each provider except vSphere is located in a separate repository. These providers use the code from the machine controller(link to MAO/machine-controller) and extend it with cloud specific logic:
+  - https://github.com/openshift/cluster-api-provider-gcp
+  - https://github.com/openshift/cluster-api-provider-azure
+  - https://github.com/openshift/cluster-api-provider-aws
+  - https://github.com/openshift/machine-api-operator/tree/master/pkg/controller/vsphere
+  - https://github.com/openshift/cluster-api-provider-openstack
+  - https://github.com/openshift/cluster-api-provider-baremetal
+  - https://github.com/openshift/cluster-api-provider-ovirt
+
+### How to start contributing
+
+The Machine API Operator and it's components are written in Golang, make sure Golang is installed in your machine and it's
+version is at least `1.13`
+
+First you need to determine if the change is related to cloud provider specific logic or it's related to all providers? Maybe you want to contribute to our e2e test suite?
+
+If it is a cloud provider related change, you should fork and clone one of provider repositories, the list can be found above.
+
+If the change is related to all providers, you should fork and clone machine-api-operator.
+
+In order to add changes to our e2e test suite fork and clone [cluster-api-actuator-pkg](https://github.com/openshift/cluster-api-actuator-pkg).
+
+Our project follows the [GitHub flow](https://guides.github.com/introduction/flow/) for contributing the code.
+
+## How to run unit tests
+You may want to run unit tests before pushing changes. It can be done in a similar way for  Machine API Operator and cloud providers repositories.
+
+Prerequisite:
+```
+git checkout github.com/openshit/$repository_name
+cd $repository_name
+```
+In order to run the unit tests locally on your machine run the following command:
+```
+NO_DOCKER=1 make test
+```
+*Note* :
+If you run this command inside the machine-api-operator directory, it will run unit tests for machine, machineset, machine health check controllers and vsphere provider.
+If this command is run inside a cloud provider repository you will run only cloud provider specific tests.
+
+## How to run a component locally for testing
+### Running machine controller
+Prerequisites:
+```
+git checkout github.com/openshit/$repository_name
+cd $repository_name
+```
+Make sure your $KUBECONFIG is set properly, because it will be used to interact with your cluster.
+
+First step is to scale down the cluster version operator.
+
+“At the heart of OpenShift is an operator called the Cluster Version Operator. This operator watches the deployments and images related to the core OpenShift services, and will prevent a user from changing these details. If I want to replace the core OpenShift services I will need to scale this operator down.” - https://notes.elmiko.dev/2020/08/18/tips-experimenting-mapi.html (Thanks for this post, Mike!)
```

Comment:
i appreciate the shoutout, but no need to thank me in the docs ;)
the reference is good enough imo